### PR TITLE
NOISSUE - add nats suport

### PR DESCRIPTION
**Pull request by mteodor**
adding nats support
agent will be able to forward messages from mqtt to local running nats
messages that arrive to
channels/<channel_id>/messages/nats/subtopic1/subtopic2 will be published to nats on topic
subtopic1.subtopic2

### CODE REVIEW

**Comment by drasko** [comment]
This specification is not quite correct.

I do agree that commands to toher microservices are sent via NATS, so that we avoid service discovery.

But in that case you need:
- To have a `CMD` or something over the command channel - i.e. you need to understand that you need to forward command to a particular microservice
- Mircoresrvies should not subscribe to `channel` topics, which is used for messages, but rather to something like `command/<microservice_unique_name>`, so that command is forwarded to them in a different namespace (do not overlap with messaging)
- Request-reply mechanism of NATS should be used most probably (to simulate RPC)

**Comment by drasko** [comment]
Additionally - NATS support should be optional. One should be capable to run Agent without NATS.

**Comment by mteodor** [comment]
when sending commands to mqtt topic
channels/<channel_id>/messages/services/<service_name>
agent will republish them locally on the nats to topic
commands.service_name

if nats is not up and running agent will just report a warning

**Comment by drasko** [comment]
@mteodor this has been merged into `nats` branch, right? If yes, please close this PR.

File: internal/pkg/mqtt/sub.go [hunk 1]
```diff
@@ -26,27 +34,65 @@ type broker struct {
...
+	last := len(natsTopic) - 1
+	if natsTopic[last] == '/' {
+		natsTopic = natsTopic[:last]
+	}
+	natsTopic = strings.ReplaceAll(natsTopic, "/", ".")
+
+	isEmpty := func(s string) bool {
+		return (len(s) == 0)
+	}
+	// filter empty array element, avoid topic topic.sub.sub...
```
Comment (drasko) [inline internal/pkg/mqtt/sub.go]: Comments start with capital letter.

File: internal/pkg/mqtt/sub.go [hunk 2]
```diff
@@ -26,27 +34,65 @@ type broker struct {
...
+	filtered := filter.Drop(strings.Split(natsTopic, "."), isEmpty).([]string)
+	natsTopic = strings.Join(filtered, ".")
+
+	return fmt.Sprintf("%s.%s", commands, natsTopic)
+
+}
+
 // handleMsg triggered when new message is received on MQTT broker
```
Comment (drasko) [inline internal/pkg/mqtt/sub.go]: Same here

File: internal/pkg/mqtt/sub.go [hunk 3]
```diff
@@ -26,27 +34,65 @@ type broker struct {
...
 
+// handleNatsMsg triggered when new message is received on MQTT broker
+func (b *broker) handleNatsMsg(mc paho.Client, msg paho.Message) {
+	topic := extractNatsTopic(msg.Topic())
+	b.nats.Publish(topic, msg.Payload())
+}
+
+func extractNatsTopic(topic string) string {
+	i := strings.LastIndex(topic, servTopic) + len(servTopic)
+	natsTopic := topic[i:]
```
Comment (drasko) [inline internal/pkg/mqtt/sub.go]: Validity of the topic can be checked via regexp, there is an example in the CoAP adapter.

This topic handling looks confusing without additional comments of what topic looks like and what you are trying to achieve.

**Pull request closed without merging**